architus/architus
# Logs backend MVP

## Summary

This pull request is designed to be the MVP for the backend event pipeline & GraphQL API of the logs service.

### Goals

- Have all services running to ingest a small subset of log events into a data store. These events are, at the moment:
  - MessageSend
  - MessageReply
  - ReactionAdd
  - ReactionRemove
  - ReactionBulkRemove
  - MemberJoin
  - MemberLeave
- Provide a simple GraphQL API for querying the log data
- Make log ingestion opt-in via the use of feature flags

### Non-goals

- Authentication/rate-limiting for the GraphQL API -- This API will not be publicly exposed to the internet and will only be usable from within the cluster for testing. Authentication/rate-limiting will be added at a later point
- Full coverage of all possible log events
- Inclusion of a history-scanning service -- this is something that will be added before product launch that will make it possible to replicate Discord audit log events into our own log events in an intelligent way that knows what periods in time it has already submitted events for.
- Nickname/color revision tracking -- this is a feature that will eventually be present for use in the web UI to show nicknames and colors in their actual point-in-time representation by tracking when they change.
- Guild uptime tracking -- this is a monitoring feature that will be present and track when the gateway ingress service is online and receiving events from Discord. It will inform the history-scanning service for potentially recovering certain events that were lost, and might also expose an API for viewing uptime history on the dashboard
- Any UI work -- the API is intended to be usable only via GraphQL clients, including a self-served "GraphQL playground" web app
- Log retention work -- we plan on implementing log retention policies and then making longer retention periods a premium feature.
- Internal authentication -- I don't have the expertise for this so I'm putting it off until after the backend MVP, but it is something we should absolutely implement before release.
- Configuration of Elasticsearch -- it's likely that we will have to do some configuration of Elasticsearch's refreshing behavior in order to strike a balance between indexing speed and data freshness. [Discord used a lazy refreshing behavior](https://blog.discord.com/how-discord-indexes-billions-of-messages-e3d5e9be866f#0059) using redis which sounds straightforward to implement, but would add a dependency on Redis that is currently not being used anywhere. More thought should be put into this at a later point in time.
- Feature-complete filtering/sorting in the GraphQL schema including full-text search
- Finalizing the GraphQL schema to allow ergonomic queries with entity variants

## Testing

To test the MVP, the following steps can be run:

1. Create the `secrets.yaml` file and populate it as needed. Of special note are all the environment variables that start with `LOGS_` and don't have a value. These include:
    - `LOGS_GATEWAY_INGRESS_CONFIG_SECRETS__DISCORD_TOKEN` and `LOGS_GATEWAY_NORMALIZE_CONFIG_SECRETS__DISCORD_TOKEN`, which both need to be populated with a valid Discord bot token
2. `tilt up -- --no-core --enable=logs` in the root of the repository
3. Have your test bot be in a guild with your Discord user for testing. Then, copy the ID of the guild
4. Use `psql` to enable the feature flag for your guild:
    1. Run `psql -h localhost -p 5432 -U autbot`, and then type the password (`autism` by default)
    2. In the Postgres REPL, run:
        1. `INSERT INTO tb_feature_flags (name,open) VALUES ('beta/logs-indexing',false);`
        2. `INSERT INTO tb_guild_features (guild_id,feature_id) VALUES (596436041045639179,1);` Be sure to replace the guild ID in this command with your own
5. Wait some duration for the feature flag to propagate. This should not take longer than a minute or two.
7. Perform one of the supported actions in your guild in a channel that the bot can see, such as adding a reaction to a message
8. Open up the GraphQL playground at http://localhost:8174/playground and execute the following query:<br /><br />
    ```graphql
    query DemoQuery {
        allEvent {
            snapshot
            pageInfo {
              currentPage
              hasPreviousPage
              hasNextPage
              itemCount
              pageCount
              perPage
              totalCount
            }
            nodes {
                id
                timestamp
                type
                guildId
                channelId
                agentId
                agentType
                agentSpecialType
                subjectId
                subjectType
                auxiliaryId
                auxiliaryType
                content
            }
        }
    }
    ```
    The response should look something like:<br /><br />
    ```json
    {
      "data": {
        "allEvent": {
          "snapshot": "bGdldl8xdnFVQzlCRVdQUmI3Vm9ZZFRJVU9abFBpRVo=",
          "pageInfo": {
            "currentPage": 0,
            "hasPreviousPage": false,
            "hasNextPage": false,
            "itemCount": 4,
            "pageCount": 1,
            "perPage": 100,
            "totalCount": 4
          },
          "nodes": [
            {
              "id": "lgev_1vkjznmHsDYYJFeB9y0CD9eCohr",
              "timestamp": "1627113245056",
              "type": "REACTION_ADD",
              "guildId": "607637793107345431",
              "channelId": "608075539629998081",
              "agentId": "74646589213257728",
              "agentType": "USER_LIKE",
              "agentSpecialType": "DEFAULT",
              "subjectId": "867594760281653259",
              "subjectType": "MESSAGE",
              "auxiliaryId": null,
              "auxiliaryType": "NONE",
              "content": "😎 :sunglasses:"
            }
          ]
        }
      }
    }
    ```
